substrate/frame/assets-vesting/src/lib.rs

```diff
@@ -0,0 +1,786 @@
+	fn add_vesting_schedule(
+		asset: AssetIdOf<T, I>,
+		who: &T::AccountId,
+		locked: BalanceOf<T, I>,
+		per_block: BalanceOf<T, I>,
+		starting_block: BlockNumberFor<T>,
+	) -> DispatchResult {
+		if locked.is_zero() {
+			return Ok(())
+		}
+
+		let vesting_schedule = VestingInfo::new(locked, per_block, starting_block);
+		// Check for `per_block` or `locked` of 0.
+		if !vesting_schedule.is_valid() {
+			return Err(Error::<T, I>::InvalidScheduleParams.into())
+		};
+
+		let mut schedules = Vesting::<T, I>::get(&asset, who).unwrap_or_default();
+
+		// NOTE: we must push the new schedule so that `exec_action`
+		// will give the correct new locked amount.
+		ensure!(schedules.try_push(vesting_schedule).is_ok(), Error::<T, I>::AtMaxVestingSchedules);
+
+		let (schedules, locked_now) =
+			Self::exec_action(schedules.to_vec(), VestingAction::Passive)?;
+
+		Self::write_vesting(asset.clone(), who, schedules)?;
+		Self::write_lock(asset, who, locked_now)?;
+
+		Ok(())
+	}
+
+	fn can_add_vesting_schedule(
+		asset: AssetIdOf<T, I>,
+		who: &T::AccountId,
+		locked: BalanceOf<T, I>,
+		per_block: BalanceOf<T, I>,
+		starting_block: BlockNumberFor<T>,
+	) -> DispatchResult {
+		// Check for `per_block` or `locked` of 0.
+		if !VestingInfo::new(locked, per_block, starting_block).is_valid() {
+			return Err(Error::<T, I>::InvalidScheduleParams.into())
+		}
+
+		ensure!(
+			(Vesting::<T, I>::decode_len(asset, who).unwrap_or_default() as u32) <
+				T::MAX_VESTING_SCHEDULES,
+			Error::<T, I>::AtMaxVestingSchedules
+		);
+
+		Ok(())
+	}
+
+	fn remove_vesting_schedule(
+		asset: AssetIdOf<T, I>,
+		who: &T::AccountId,
+		schedule_index: u32,
+	) -> DispatchResult {
+		let schedules = Vesting::<T, I>::get(&asset, who).ok_or(Error::<T, I>::NotVesting)?;
+		let remove_action = VestingAction::Remove { index: schedule_index as usize };
+
+		let (schedules, locked_now) = Self::exec_action(schedules.to_vec(), remove_action)?;
+
+		Self::write_vesting(asset.clone(), who, schedules)?;
+		Self::write_lock(asset, who, locked_now)?;
+		Ok(())
+	}
+}
+
+impl<T: Config<I>, I: 'static> VestedTransfer<T::AccountId> for Pallet<T, I>
+where
+	BalanceOf<T, I>: MaybeSerializeDeserialize + Debug,
+{
+	type Moment = BlockNumberFor<T>;
+	type AssetId = AssetIdOf<T, I>;
+	type Balance = BalanceOf<T, I>;
+
+	fn vested_transfer(
+		asset: AssetIdOf<T, I>,
+		source: &T::AccountId,
+		target: &T::AccountId,
+		locked: BalanceOf<T, I>,
+		per_block: BalanceOf<T, I>,
+		starting_block: BlockNumberFor<T>,
+	) -> DispatchResult {
+		use storage::{with_transaction, TransactionOutcome};
+		let schedule = VestingInfo::new(locked, per_block, starting_block);
+		with_transaction(|| -> TransactionOutcome<DispatchResult> {
+			let result = Self::do_vested_transfer(asset, source, target, schedule);
+
+			match &result {
+				Ok(()) => TransactionOutcome::Commit(result),
+				_ => TransactionOutcome::Rollback(result),
+			}
+		})
```

Comment:
Just a minor refactor:
```suggestion
		use storage::with_storage_layer;
		let schedule = VestingInfo::new(locked, per_block, starting_block);
		with_storage_layer(|| Self::do_vested_transfer(asset, source, target, schedule))
```